pkg/parse/root_test.go

```diff
@@ -451,6 +454,221 @@ func TestRoot_Parse(t *testing.T) {
+			discoveryClient: syncertest.NewDiscoveryClientWithErrors(true, kinds.Namespace(), kinds.Role()),
+			expectError:     true,
+			parsed: []ast.FileObject{
+				fake.Role(core.Namespace("foo")),
+				// add a faked obect in parser.parsed without CRD so it's scope will be unknown when validating
+				fake.Unstructured(kinds.Anvil(), core.Name("deploy")),
+			},
+			want: []ast.FileObject{},
+		},
+		{
+			// unknown scoped object get skipped when sending to applier
+			name:            "unknown scoped object without discovery failure",
+			format:          filesystem.SourceFormatUnstructured,
+			discoveryClient: syncertest.NewDiscoveryClientWithErrors(false, kinds.Namespace(), kinds.Role()),
+			expectError:     true,
+			parsed: []ast.FileObject{
+				fake.Role(core.Namespace("foo")),
+				// add a faked obect in parser.parsed without CRD so it's scope will be unknown when validating
+				fake.Unstructured(kinds.Anvil(), core.Name("deploy")),
+			},
+			want: []ast.FileObject{
+				fake.UnstructuredAtPath(kinds.Namespace(),
+					"",
+					core.Name("foo"),
+					core.Label(metadata.ManagedByKey, metadata.ManagedByValue),
+					core.Annotation(common.LifecycleDeleteAnnotation, common.PreventDeletion),
+					core.Annotation(metadata.ResourceManagementKey, metadata.ResourceManagementEnabled),
+					core.Annotation(metadata.GitContextKey, nilGitContext),
+					core.Annotation(metadata.SyncTokenAnnotationKey, ""),
+					core.Annotation(metadata.OwningInventoryKey, applier.InventoryID(rootSyncName, configmanagement.ControllerNamespace)),
+					core.Annotation(metadata.ResourceIDKey, "_namespace_foo"),
+					difftest.ManagedBy(declared.RootReconciler, rootSyncName),
+				),
+				fake.Role(core.Namespace("foo"),
+					core.Label(metadata.ManagedByKey, metadata.ManagedByValue),
+					core.Label(metadata.DeclaredVersionLabel, "v1"),
+					core.Annotation(metadata.DeclaredFieldsKey, `{"f:metadata":{"f:annotations":{},"f:labels":{}},"f:rules":{}}`),
+					core.Annotation(metadata.SourcePathAnnotationKey, "namespaces/foo/role.yaml"),
+					core.Annotation(metadata.ResourceManagementKey, metadata.ResourceManagementEnabled),
+					core.Annotation(metadata.GitContextKey, nilGitContext),
+					core.Annotation(metadata.SyncTokenAnnotationKey, ""),
+					core.Annotation(metadata.OwningInventoryKey, applier.InventoryID(rootSyncName, configmanagement.ControllerNamespace)),
+					core.Annotation(metadata.ResourceIDKey, "rbac.authorization.k8s.io_role_foo_default-name"),
+					difftest.ManagedBy(declared.RootReconciler, rootSyncName),
+				),
+			},
+		},
+		{
+			// happy path condition
+			name:            "known scoped object without discovery failure",
+			format:          filesystem.SourceFormatUnstructured,
+			discoveryClient: syncertest.NewDiscoveryClientWithErrors(false, kinds.Namespace(), kinds.Role()),
+			expectError:     false,
+			parsed: []ast.FileObject{
+				fake.Role(core.Namespace("foo")),
+				fakeCRD(core.Name("anvils.acme.com")),
```

Comment:
I am a little confused here. The test case name says `known scoped object ...`, but `fakeCRD` defined a CRD whose scope is unknown.

Reply:
The `fakeCRD` generates a CRD with group `acme.com` so that the `fake.Unstructured(kinds.Anvil())` which is also in tc.parsed can have a known scope, previous cases are not using fakeCRD to force the `fake.Unstructured(kinds.Anvil())` to be unknown, then get skipped when doing `splitObjects` in `cache.go`